src/lib/worker-runner-protocol.js

```diff
@@ -0,0 +1,108 @@
+const {EventEmitter} = require('events');
+const split2 = require('split2');
+
+/**
+ * This is an implementation of the worker-runner protocol documented at
+ * https://github.com/taskcluster/taskcluster-worker-runner/blob/master/protocol.md
+ */
+
+/**
+ * A transport should have a `send(message)` method to send messages,
+ * and should emit a `message` event when one is received.  Since this
+ * implements only the worker side of the protocol, invalid lines are
+ * simply ignored.
+ *
+ * StreamTransport implements this interface using Node streams.
+ */
+class StreamTransport extends EventEmitter {
+  constructor(input, output) {
+    super();
+
+    // line-buffer the input and react to individual messages
+    const lines = input.pipe(split2());
+
+    lines.on('data', line => {
+      if (!line.startsWith('~{') || !line.endsWith('}')) {
+        return;
+      }
+      let msg;
+      try {
```

Comment:
iiuc, we should validate the message starts with `~`, no?

Reply:
Maybe, but I'm not sure what to do with it otherwise -- until this lands, docker-worker has just ignored its stdin.  I'll add that.

Reply:
Oh, now that I look, it's validated a few lines earlier:
https://github.com/djmitche/docker-worker/blob/086fab4e1981c50567c3a94577b4b80a126844bb/src/lib/worker-runner-protocol.js#L25-L27
```js
      if (!line.startsWith('~{') || !line.endsWith('}')) {
        return;
      }
```